### PR TITLE
recipe2plan: Explicit error handing when mapping to volatile create handles (stores). 

### PR DESCRIPTION
**Pull request by alxrsngrtn**
Instead of getting vague errors along the lines of "Cannot find associated store", this PR provides an up-front check against mapping or copying a handle from a volatile create handle (and thus, a volatile store). 

In addition, I added a small fix to the complex recipe test case; now it can produce a plan.

### CODE REVIEW

File: src/tools/storage-key-recipe-resolver.ts [hunk 1]
```diff
@@ -172,3 +182,13 @@ export function findLongRunningArcId(recipe: Recipe): string | null {
   }
   return null;
 }
+
+
+/** Determines if create handle is volatile. */
+function isVolatileCreateHandle(handle: Handle): boolean {
+  if (handle.capabilities.isEmpty()) return true;
+  if (!handle.capabilities.isPersistent &&
```
Comment (alxrsngrtn) [inline src/tools/storage-key-recipe-resolver.ts:190]: Question for the reviewer: Can handles be volatiles and queryable?

**Comment by alxrsngrtn** [comment]
A quick aside: I made formatting changes. You'll have an easier time reviewing this change by ignoring whitespace changes in the diff.

File: src/tools/storage-key-recipe-resolver.ts (hunk 1 above)
Comment (mmandlis) [reply src/tools/storage-key-recipe-resolver.ts:190]: not at the moment, but we are rethinking capabilities, so in the future stores will be able to be in-memory and queryable.

File: src/tools/storage-key-recipe-resolver.ts [hunk 2]
```diff
@@ -172,3 +182,13 @@ export function findLongRunningArcId(recipe: Recipe): string | null {
...
+    !handle.capabilities.isTiedToRuntime &&
+    !handle.capabilities.isTiedToArc) return true;
```
Comment (mmandlis) [inline src/tools/storage-key-recipe-resolver.ts:192]: this seems wrong, TiedToArc does map to volatile right now

File: src/tools/tests/storage-key-recipe-resolver-test.ts [hunk 3]
```diff
@@ -374,8 +374,7 @@ describe('recipe2plan', () => {
 
       const resolver = new StorageKeyRecipeResolver(manifest);
       for (const it of (await resolver.resolve())) {
-        assert.isTrue(it.isResolved());
-      }
+        assert.isTrue(it.isResolved()); }
```
Comment (mmandlis) [inline src/tools/tests/storage-key-recipe-resolver-test.ts:377]: is this change intentional? seems wrong, `}` should go on the next line

File: src/tools/tests/storage-key-recipe-resolver-test.ts [hunk 4]
```diff
@@ -393,5 +392,33 @@ describe('recipe2plan', () => {
...
+      const manifest = await Manifest.parse(`\
+    particle Reader
+      data: reads Thing {name: Text}
+
+    particle Writer
+       data: writes Thing {name: Text}
+    
+    @trigger
+      launch startup
+      arcId writeArcId
```
Comment (mmandlis) [inline src/tools/tests/storage-key-recipe-resolver-test.ts:405]: just as a side note (no change needed in this PR), this behavior is temporary because lack of `persistent` capability maps directly to `volatile` in the current implementation.
we would change it soon, so that if a recipe is mapping the store, it would become `ramdisk`, but not `volatile` (basically allow mapping of an in-memory store).
just an fyi

File: src/tools/tests/storage-key-recipe-resolver-test.ts (hunk 3 above)
Comment (alxrsngrtn) [reply src/tools/tests/storage-key-recipe-resolver-test.ts:377]: I applied an auto-formatter; I'm surprised it did such a weird thing.

File: src/tools/tests/storage-key-recipe-resolver-test.ts (hunk 4 above)
Comment (alxrsngrtn) [reply src/tools/tests/storage-key-recipe-resolver-test.ts:405]: Ah, thanks for that context. Yeah, that would also address the underlying issue here. 
The change here is easily reversible, when we're ready to make the switch.

File: src/tools/storage-key-recipe-resolver.ts (hunk 2 above)
Comment (alxrsngrtn) [reply src/tools/storage-key-recipe-resolver.ts:192]: Given your above comment, I'm going to s/isTiedToArc/isQueryable/, does that sound right to you?

File: src/tools/storage-key-recipe-resolver.ts [hunk 5]
```diff
@@ -148,8 +151,15 @@ export class StorageKeyRecipeResolver {
             `Handle ${handle.localName} mapped to ephemeral handle '${match.id}'.`
           );
         }
+
+        if (isVolatileCreateHandle(match)) {
```
Comment (mmandlis) [inline src/tools/storage-key-recipe-resolver.ts:155]: sorry for the late insight. could you check `match.storageKey instanceOf VolatileStorageKey`, instead of having a helper method looking at capabilities?

Comment (alxrsngrtn) [reply src/tools/storage-key-recipe-resolver.ts:155]: I tried that in this PR before. Unfortunately, it's not guaranteed that `match` has a `storageKey` set. If it did, I would re-use the `isVolatileStore` implementation in Manifest.ts (with some minor modification).

Comment (mmandlis) [reply src/tools/storage-key-recipe-resolver.ts:155]: if it doesn't have a `storageKey` doesn't it mean that it's a handle from a non long-running arc? so it cannot be mapped, but for a different reason.

Comment (alxrsngrtn) [reply src/tools/storage-key-recipe-resolver.ts:155]: I'll do my best to explain my reasoning here, it might be better to VC. 

It doesn't have a storage key at this time since we create the storage keys later in this module. First we validate (this step), then we resolve the recipe, then we create the keys for create handles. We do these steps in a loop for each recipe. 

Before this check, we'd get an error at the resolution step. Adding this to the validation offers cleaner error messaging. 

I tried a solution that was dependent on the storage keys being present, but it was complicated by the existing order. Changing the order of processing would be unraveling. I tried null-checking the keys, but then that makes the system dependent on the order of recipes, which is fragile.

Comment (mmandlis) [reply src/tools/storage-key-recipe-resolver.ts:155]: i see. 

this approach to checking whether handle is mappable or not is very limited (volatile is not the only reason resolution might fail) and hacky  (reasoning about volatile in general is not absolutely correct atm), I'd rather we don't do it.
if you remove this check, the test will fail with `cannot find associated store` (thrown in recipe-resolver). I think if you were to add the `id` of the store that couldn't be found, that would significantly improve the error message without compromising code quality. in general, it should be recipe-resolver's job to report a comprehensive error if resolution fails.

Comment (alxrsngrtn) [reply src/tools/storage-key-recipe-resolver.ts:155]: I'm happy with going this direction. I'll get on it now.

**Comment by alxrsngrtn** [comment]
I'm planning on merging despite cloud build's failure -- it seems to be caused by a flaky test.

**Comment by alxrsngrtn** [comment]
Closed in favor of #5164
